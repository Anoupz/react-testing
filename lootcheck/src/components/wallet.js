import React, {Component} from 'react';
import {Segment, Button} from 'semantic-ui-react';
import {connect} from 'react-redux';
import {deposit, withDraw} from '../actions/balance';

export class Wallet extends Component {
  constructor(props) {
    super(props);
    this.state = {
      balance: undefined
    }
  }

  updateBalance = event => {
    this.setState({balance: parseInt(event.target.value, 10)});
  };

  makeDeposit = () => {
    this.props.deposit(this.state.balance);
  };

  withDrawFrom = () => {
    this.props.withDraw(this.state.balance);
  };

  render() {
    return (
      <Segment basic>
        <h3 className="balance">Wallet Balance: {this.props.balance}</h3>
        <br/>
        <input type="text" className="input-wallet" onChange={this.updateBalance}/>
        <Button className="btn-deposit" onClick={this.makeDeposit}>Deposit</Button>
        <Button className="btn-withdraw" onClick={this.withDrawFrom}>WithDraw</Button>
      </Segment>
    )
  }
}

export default connect(state => {
  return {balance: state.balance};
}, {deposit, withDraw})(Wallet);