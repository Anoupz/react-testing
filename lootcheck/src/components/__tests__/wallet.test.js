import React from "react";
import {shallow} from "enzyme";
import {Wallet} from '../wallet';

describe('Wallet', () => {
  const mockDeposit = jest.fn();
  const mockWithDraw = jest.fn();
  const props = {balance: 20, deposit: mockDeposit, withDraw: mockWithDraw};
  const wallet = shallow(<Wallet {...props}/>);

  it('renders correctly', () => {
    expect(wallet).toMatchSnapshot();
  });

  it('displays the balance from props', () => {
    expect(wallet.find('.balance').text()).toEqual('Wallet Balance: 20');
  });

  it('creates an input to deposit into or withdraw from balance', () => {
    expect(wallet.find('.input-wallet').exists()).toBeTruthy();
  });

  describe('when the user types into wallet input', () => {
    const userBalance = '25';
    beforeEach(() => {
      wallet.find('.input-wallet').simulate('change', {target: {value: userBalance}});
    });

    it('updates the local wallet balance in the state and it converts to a number', () => {
      expect(wallet.state().balance).toEqual(parseInt(userBalance, 10));
    });

    describe('user wants to make a deposit', () => {

      beforeEach(() => {
        wallet.find('.btn-deposit').simulate('click')
      });

      it('dispatches the deposit() it receives from props with local balance', () => {
        expect(mockDeposit).toHaveBeenCalledWith(parseInt(userBalance, 10));
      });
    });

    describe('user wants to make a withdrawal', () => {

      beforeEach(() => {
        wallet.find('.btn-withdraw').simulate('click')
      });

      it('dispatches the withdraw() it receives from props with local balance', () => {
        expect(mockWithDraw).toHaveBeenCalledWith(parseInt(userBalance, 10));
      });
    });
  });

});