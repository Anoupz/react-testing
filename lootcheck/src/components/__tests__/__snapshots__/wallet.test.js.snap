// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Wallet renders correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Wallet
    balance={20}
    deposit={[Function]}
    withDraw={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": null,
    "nodeType": "function",
    "props": Object {
      "basic": true,
      "children": Array [
        <h3
          className="balance"
>
          Wallet Balance: 
          20
</h3>,
        <br />,
        <input
          className="input-wallet"
          onChange={[Function]}
          type="text"
/>,
        <Button
          as="button"
          className="btn-deposit"
          onClick={[Function]}
>
          Deposit
</Button>,
        <Button
          as="button"
          className="btn-withdraw"
          onClick={[Function]}
>
          WithDraw
</Button>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": null,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "Wallet Balance: ",
            20,
          ],
          "className": "balance",
        },
        "ref": null,
        "rendered": Array [
          "Wallet Balance: ",
          20,
        ],
        "type": "h3",
      },
      Object {
        "instance": null,
        "key": null,
        "nodeType": "host",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": "br",
      },
      Object {
        "instance": null,
        "key": null,
        "nodeType": "host",
        "props": Object {
          "className": "input-wallet",
          "onChange": [Function],
          "type": "text",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": null,
        "nodeType": "class",
        "props": Object {
          "as": "button",
          "children": "Deposit",
          "className": "btn-deposit",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": "Deposit",
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": null,
        "nodeType": "class",
        "props": Object {
          "as": "button",
          "children": "WithDraw",
          "className": "btn-withdraw",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": "WithDraw",
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": null,
      "nodeType": "function",
      "props": Object {
        "basic": true,
        "children": Array [
          <h3
            className="balance"
>
            Wallet Balance: 
            20
</h3>,
          <br />,
          <input
            className="input-wallet"
            onChange={[Function]}
            type="text"
/>,
          <Button
            as="button"
            className="btn-deposit"
            onClick={[Function]}
>
            Deposit
</Button>,
          <Button
            as="button"
            className="btn-withdraw"
            onClick={[Function]}
>
            WithDraw
</Button>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": null,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "Wallet Balance: ",
              20,
            ],
            "className": "balance",
          },
          "ref": null,
          "rendered": Array [
            "Wallet Balance: ",
            20,
          ],
          "type": "h3",
        },
        Object {
          "instance": null,
          "key": null,
          "nodeType": "host",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": "br",
        },
        Object {
          "instance": null,
          "key": null,
          "nodeType": "host",
          "props": Object {
            "className": "input-wallet",
            "onChange": [Function],
            "type": "text",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": null,
          "nodeType": "class",
          "props": Object {
            "as": "button",
            "children": "Deposit",
            "className": "btn-deposit",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "Deposit",
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": null,
          "nodeType": "class",
          "props": Object {
            "as": "button",
            "children": "WithDraw",
            "className": "btn-withdraw",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "WithDraw",
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "disableLifecycleMethods": true,
  },
}
`;
